From 9bc5e6cecf1f919fe16da3689d76bcc39c23ee4a Mon Sep 17 00:00:00 2001
From: Roy ZANG <tie-fei.zang@freescale.com>
Date: Wed, 10 Oct 2012 09:40:48 +0000
Subject: [PATCH 096/227] Powerpc/PCIe fix the link judgment issue

This patch fixes the link judgment issue introduced by commit:
Author: Roy ZANG <tie-fei.zang@freescale.com>
Date:   Wed Sep 26 04:23:56 2012 +0000

    T4/PCIe: use status/control register to judge the link status

    PCIe IP 3.0, which has status/control register, use status/control
    register to judge the link status

    Signed-off-by: Roy Zang <tie-fei.zang@freescale.com>
    Signed-off-by: Andy Fleming <afleming@freescale.com>

For PCIe IP version lower than 3.0, still use the original link
judgment method.

Signed-off-by: Roy Zang <tie-fei.zang@freescale.com>
Signed-off-by: Andy Fleming <afleming@freescale.com>
[Kevin: Original patch taken from fsl sdk 1.3.1
QorIQ-SDK-V1.3.1-SOURCE-20121220-yocto.iso.]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 arch/powerpc/sysdev/fsl_pci.c |   21 ++++++++++++---------
 arch/powerpc/sysdev/fsl_pci.h |    1 +
 2 files changed, 13 insertions(+), 9 deletions(-)

diff --git a/arch/powerpc/sysdev/fsl_pci.c b/arch/powerpc/sysdev/fsl_pci.c
index af882a2..e8fadb9 100644
--- a/arch/powerpc/sysdev/fsl_pci.c
+++ b/arch/powerpc/sysdev/fsl_pci.c
@@ -78,23 +78,26 @@ static int __init fsl_pcie_check_link(struct pci_controller *hose,
 		pr_debug("PCI memory map start 0x%016llx, size 0x%016llx\n",
 		    (u64)rsrc->start, (u64)rsrc->end - (u64)rsrc->start + 1);
 		pci = ioremap(rsrc->start, rsrc->end - rsrc->start + 1);
-		if (pci) {
+		if (!pci) {
+			dev_err(hose->parent, "Unable to map PCIe registers\n");
+			return -ENOMEM;
+		}
+		if (in_be32(&pci->block_rev1) >= PCIE_IP_REV_3_0) {
 #define PEX_CSR0_LTSSM_MASK    0xFC
 #define PEX_CSR0_LTSSM_SHIFT    2
 			val = (in_be32(&pci->pex_csr0)
 				& PEX_CSR0_LTSSM_MASK) >> PEX_CSR0_LTSSM_SHIFT;
-			iounmap(pci);
 			if (val != 0x11)
 				return 1;
-		} else {
-			dev_err(hose->parent, "Unable to map PCIe registers\n");
-			return -ENOMEM;
+			iounmap(pci);
+			return 0;
 		}
-	} else {
-		early_read_config_dword(hose, 0, 0, PCIE_LTSSM, &val);
-		if (val < PCIE_LTSSM_L0)
-			return 1;
+		iounmap(pci);
 	}
+	early_read_config_dword(hose, 0, 0, PCIE_LTSSM, &val);
+	if (val < PCIE_LTSSM_L0)
+		return 1;
+
 	return 0;
 }
 
diff --git a/arch/powerpc/sysdev/fsl_pci.h b/arch/powerpc/sysdev/fsl_pci.h
index 3dc22e1..5be4eb9 100644
--- a/arch/powerpc/sysdev/fsl_pci.h
+++ b/arch/powerpc/sysdev/fsl_pci.h
@@ -19,6 +19,7 @@
 #define PCIE_LTSSM	0x0404		/* PCIE Link Training and Status */
 #define PCIE_LTSSM_L0	0x16		/* L0 state */
 #define PCIE_IP_REV_2_2		0x02080202 /* PCIE IP block version Rev2.2 */
+#define PCIE_IP_REV_3_0		0x02080300 /* PCIE IP block version Rev3.0 */
 #define PIWAR_EN		0x80000000	/* Enable */
 #define PIWAR_PF		0x20000000	/* prefetch */
 #define PIWAR_TGI_LOCAL		0x00f00000	/* target - local memory */
-- 
1.7.9.7

